--- aclocal.m4.old	2007-04-20 14:54:50.000000000 -0700
+++ aclocal.m4	2007-04-20 13:03:19.000000000 -0700
@@ -2855,7 +2855,7 @@
   # Commands to make compiler produce verbose output that lists
   # what "hidden" libraries, object files and flags are used when
   # linking a shared library.
-  output_verbose_link_cmd='$CC -shared $CFLAGS -v conftest.$objext 2>&1 | grep "\-L"'
+  output_verbose_link_cmd='$CC -shared $CFLAGS -v conftest.$objext 2>&1 | grep "\-L" | tail -n 1'
 
 else
   GXX=no
